Okta-Source: Add Parameter sub domain to accept other domains than okta.com

haroldsnyers opened the pull request:
## What
The current Okta connector has the otka domain hardcoded as okta.com. In some cases, we might need to be able to sync data from other domain like oktapreview.com or custom ones. 

Related to this issue : https://github.com/airbytehq/airbyte/issues/46375

Example : 
baseurl: myorg.okta.com
sub domain : myorg
domain: okta.com

## How
Before : 
baseurl: https://myorg.okta.com
domain : myorg

Solution : 
baseurl: https://myorg.okta.com
sub domain : myorg
domain: okta.com

Also upgrade to latest airbyte sdk as this source was still on the sdk 0.90.0.

## User Impact
Migration function has been updated to demand migration if the fields `sub_domain` and `domain` are not part of the config.
New users will be able to set both the sub domain and domain. 

## Can this PR be safely reverted and rolled back?
- [ ] YES 💚
- [ ] NO ❌

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=46540) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=46540) before we can accept your contribution.<br/><hr/>**haroldsnyers** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=46540) it.</sub>

evyros commented:
@haroldsnyers Great contribution, definitely needed.
I would simplify it by expecting the `domain` parameter to include the subdomain instead of splitting it to 2 parameters.
Let me know what you think.

haroldsnyers commented:
@evyros Hey, thanks
That would work. but I am a bit afraid on how we can check if some migration is needed. 
Also, I follow this pattern given what I saw in other ingestion tools like Fivetran where they use also `sub_domain` and `domain`

natikgadzhi commented:
Thank you for this! Especially for attempting to bump the CDK while you're there. Kicked-off the CI, looking into the code now.

natikgadzhi commented on airbyte-integrations/connectors/source-okta/pyproject.toml:
```diff
@@ -16,7 +16,7 @@ packages = [ { include = "source_okta" }, {include = "main.py"} ]
 [tool.poetry.dependencies]
 python = "^3.9,<3.12"
-airbyte-cdk = "^0"
+airbyte-cdk = "^3"
```
Did you test this and verified that anything above 4.0 breaks, or is this because you considered CDK 3.* current?

There are a lot of performance gains in CDK 5, I'd suggest attempting 5.10.2+.

```suggestion
airbyte-cdk = "^5.11"
```

natikgadzhi replied:
If you do choose to accept this, you would need to re-run `poetry lock` locally and push that file.

natikgadzhi commented on airbyte-integrations/connectors/source-okta/source_okta/components.py:
```diff
@@ -52,6 +55,7 @@ def get_refresh_request_body(self) -> Mapping[str, Any]:
         }
 
     def refresh_access_token(self) -> Tuple[str, int]:
+        print(self.token_refresh_endpoint)
```
If you do have logging here, you probably don't want a rogue print statement. Use logging all the way?

natikgadzhi commented on airbyte-integrations/connectors/source-okta/source_okta/components.py:
```diff
@@ -90,15 +95,15 @@ def token(self) -> str:
             "iss": client_id,
             "sub": client_id,
-            "aud": f"https://{domain}.okta.com/oauth2/v1/token",
+            "aud": f"https://{sub_domain}.{domain}/oauth2/v1/token",
```
I see what you did there, but effectively you're changing what `domain` config key is doing. I agree that the naming that you proposed works better. I also am not sure how config migrations would work.

Instead of effectively renaming a key, what you could do instead is an additive config change with a default value for the new setting:
- Use `config['domain']` for the subdomain, as before
- Make a new thing called `config['environment_domain']` that can either be a string with default value of `okta.com`, or a `oneOf` that covers `okta.com`, `oktapreview.com` etc.

That way, all connections would be able to update to this version without breaking their syncs.

natikgadzhi commented on airbyte-integrations/connectors/source-okta/source_okta/config_migration.py:
```diff
@@ -31,11 +31,12 @@ def should_migrate(cls, config: Mapping[str, Any]) -> bool:
             > Raises the Exception if the structure could not be migrated.
         """
-        return "domain" not in config
+        return "sub_domain" not in config and "domain" not in config
```
This is murky territory. What if `domain` is in config, but `sudbomain` is not, and what if the domain setting is "myorg". Wouldn't you try using `something.myorg.com` as the API endpoint in that case unless it's migrated? 

My vote is to stick with how those names were previously used and add a new config input name. Naming will be wonky, but you avoid a breaking change that way.

natikgadzhi reviewed (changes requested):
I'd take a different approach to the config inputs and add a new one for the environment instead of changing what the domain one did before. 

That would scope your change in a way that's easier to migrate to.

haroldsnyers commented:
Hey @natikgadzhi I have updated the code given your comments, I like your approach for a more friendly migration. I have also updated the sdk to your suggested version. 

Also, i am not sure why but CLA doesn't seem to see my commits as commits for a github user 
![image](https://github.com/user-attachments/assets/663cb3a0-88cd-4158-8dea-46e340e1ab95)

haroldsnyers commented:
@evyros with the new sdk, and how it emits messages, I am struggling to make the test_migration unit test work, something to do with the capsys

```
=================================== FAILURES ===================================
│ │         ┃ ____________________ TestMigrateConfig.test_migrate_config _____________________
│ │         ┃
│ │         ┃ self = <test_migration.TestMigrateConfig object at 0xffff6d6a0d30>
│ │         ┃ capsys = <_pytest.capture.CaptureFixture object at 0xffff6d6a0130>
│ │         ┃
│ │         ┃     def test_migrate_config(self, capsys):
│ │         ┃         config = load_config(self.test_not_migrated_config_path)
│ │         ┃         assert "domain" not in config
│ │         ┃         migration_instance = OktaConfigMigration()
│ │         ┃         migration_instance.migrate([CMD, "--config", self.test_not_migrated_config_path], SOURCE)
│ │         ┃         captured = capsys.readouterr()
│ │         ┃         logger.info(f"Captured Output: {captured.out}")  # Add this to see what's actually captured
│ │         ┃         control_msg = json.loads(captured.out)
│ │         ┃ >       assert control_msg["type"] == Type.CONTROL.value
│ │         ┃ E       TypeError: 'NoneType' object is not subscriptable
│ │         ┃
│ │         ┃ unit_tests/test_migration.py:38: TypeError
```

natikgadzhi commented:
Hm. What's `captured_out` if you print it? Seems like it wasn't emitted correctly, OR it's not JSON and could not be parsed.

It could be ok to drop the config migration tests, as long as regression tests would work, which I can kick off.

natikgadzhi commented on airbyte-integrations/connectors/source-okta/unit_tests/test_streams.py:
```diff
@@ -22,67 +18,18 @@ def get_stream_by_name(stream_name: str, config: Mapping[str, Any]) -> Stream:
         inputs = {"stream_slice": None, "stream_state": None, "next_page_token": None}
         assert list(stream.retriever.requester.get_request_params(**inputs).keys())[0] == "since"
 
-    def test_incremental_okta_stream_backoff_time(self, oauth_config, url_base, start_date):
```
Why delete a bunch of these tests?

haroldsnyers replied:
the reason I deleted those is that I was not able to find in the new sdk, the equivalent methods used such as the `should_retry`, or the `backoff_time`

So as I looked at the sdk code, the oktaSource declarative syntax uses As simpleRetriever with a Requester. I was not able to find equivalent methods

example
`assert stream.retriever.requester._should_retry(response_mock) == should_retry`
